File: Java/ex15.java
class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(String mensagem) {
        super(mensagem);
    }
}

class ContaBancaria {
    private double saldo;
    private String titular;

    public ContaBancaria(String titular, double saldoInicial) {
        this.titular = titular;
        this.saldo = saldoInicial;
    }

    public void depositar(double valor) {
        saldo += valor;
    }

    public void sacar(double valor) throws SaldoInsuficienteException {
        if (valor > saldo) {
            throw new SaldoInsuficienteException("Saldo insuficiente para realizar o saque.");
        }
        saldo -= valor;
    }

    public double getSaldo() {
        return saldo;
    }

    public String getTitular() {
        return titular;
    }
}

public class Main {
    public static void main(String[] args) {
        ContaBancaria conta = new ContaBancaria("João", 1000.00);

        try {
            conta.sacar(1500.00);
        } catch (SaldoInsuficienteException e) {
            System.out.println(e.getMessage());
        }

        conta.depositar(500.00);
        try {
            conta.sacar(1500.00);
        } catch (SaldoInsuficienteException e) {
            System.out.println(e.getMessage());
        }

        System.out.println("Saldo atual: R$ " + conta.getSaldo());
    }
}

File: python/ex15.py

class SaldoInsuficienteException(Exception):
    def __init__(self, mensagem="Saldo insuficiente para a operação."):
        self.mensagem = mensagem
        super().__init__(self.mensagem)


class ContaBancaria:
    def __init__(self, titular, saldo_inicial=0):
        self.titular = titular
        self.saldo = saldo_inicial

    def depositar(self, valor):
        self.saldo += valor
        print(f"Depósito de R${valor:.2f} realizado com sucesso. Saldo atual: R${self.saldo:.2f}")

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException(f"Tentativa de saque: R${valor:.2f}. Saldo disponível: R${self.saldo:.2f}")
        self.saldo -= valor
        print(f"Saque de R${valor:.2f} realizado com sucesso. Saldo atual: R${self.saldo:.2f}")


def main():
    conta = ContaBancaria("João", 100)

    try:
        conta.sacar(150)  
    except SaldoInsuficienteException as e:
        print(e)

    conta.depositar(50)  
    try:
        conta.sacar(100)  
    except SaldoInsuficienteException as e:
        print(e)

if __name__ == "__main__":
    main()
